espresso: Add real devices documentation (#404)
saucelabs/sauce-docs: docs/testrunner-toolkit/configuration/espresso.md

    - "10.0"
```

## `devices`

> __Description__: Field for requesting a real device for the test.

Reviewer: ```suggestion
__Description__: Field for requesting real devices for the test.
```

You can request a specific device using its ID, or you can specify a set of criteria to choose the first available device that matches the specifications.

When an ID is specified, it supersedes the other settings.

__Type__: *[]object*

__Example__:
```yaml
devices:
  - name: "Google Pixel.*"
    platformVersion: 8.1
  - id: Google_Pixel_2_real_us
```

### `id`

__Description__: Request a specific device by its id. You can look up device IDs in the Sauce Labs app or using our [Get Devices API request](https://docs.saucelabs.com/dev/api/rdc#get-devices).

__Type__:  *string*

__Example__:

```yaml
        id: Google_Pixel_2_real_us
```

### `name`

__Description__:  Request a device by its name.

__Type__:  *string*

__Example__:

Using complete name:
```yaml
      - name: Google Pixel 4 XL
```

Using pattern matching:
```yaml
        name: Google Pixel.*
```

### `platformVersion`

> __Description__: Request that the device matches a specific platform version.

Reviewer: ```suggestion
__Description__: Request that the device matches a specific platform version. You can use the `*` wildcard to denote flexibility.
```

__Type__:  *string*

__Example__:

```yaml

>         platformVersion: 8.1

Reviewer: ```suggestion
        platformVersion: 8.*
```

    - "10.0"
```

## `devices`

__Description__: Field for defining devices request.

A real device can be selected using two different manner:
- Specific device selection, through device IDs
- Filtered device selection, with several contraints such as `name`, `platformVersion`, etc...

> When an ID is specified, it is prioritary on other settings.

Author: If the device is not available, it will wait until it become available. Specifying an `id` involves a strict match on that `id`.
No timelimit that I am aware of. We will introduce a limit at `saucectl` level later, that will just make `saucectl` stop waiting when a specified duration has been reached. The job itself will still be waiting for the device to be available.

__Type__: *[]object*

__Example__:
```yaml
devices:
  - name: "Google Pixel.*"
    platformVersion: 8.1

>   - id: Google_Pixel_2_real_us

Author: In that example it's 2 different devices, but may be we should split it more, so the difference is more obvious.
Do you have any suggestion on how to make that more clear ?

>   - id: Google_Pixel_2_real_us

Reviewer: Oh. I see. I missed the `-` in front of the `id`. It is clear as is, but I did not realize you could choose multiple real devices for a single test suite. In that case, I will update the "devices" description. Sorry about that.

>   - id: Google_Pixel_2_real_us

Author: No problem at all. Those `-` may be sneaky 😄

```

### `id`

> __Description__: Request a specific device by its id.

Author: We definitively should ! It's much more convenient this way to the user.

> __Description__: Request a specific device by its id.

Reviewer: ```suggestion
__Description__: Request a specific device by its id. You can look up device IDs in the Sauce Labs app or using our [Get Devices API request](https://docs.saucelabs.com/dev/api/rdc#get-devices).
```

__Type__:  *string*

__Example__:

```yaml
        id: Google_Pixel_2_real_us
```

### `name`

__Description__:  Request a device by its name.

__Type__:  *string*

__Example__:

Using complete name:
```yaml
      - name: Google Pixel 4 XL
```

Using pattern matching:
```yaml
        name: Google Pixel.*
```

### `platformVersion`

> __Description__: Request a device with a specific platform version.

Author: @sweeneyskirt-sl missed some comments.
`Request any device matching` let me though that it will be executed one per matching device.

Is it ok to replace any by `that the` ?
[Like here](https://github.com/saucelabs/sauce-docs/pull/404/files#r640006714)

> __Description__: Request a device with a specific platform version.

Reviewer: Definitely. I agree that is clearer. :)

__Type__:  *string*

__Example__:

```yaml

>         platformVersion: 8.1

Author: I don't think. I'll confirm that with service team.

>         platformVersion: 8.1

Author: It is supported !

>         platformVersion: 8.1

Reviewer: I thought so! Cool. Let's reflect that in the description.

```

> ### `options`

Author: Technically you can request a device without any constraints (no `id`, no `name` and no `options`) from `saucectl` and `rdc` perspective. I don't know how likely a user will do that as you don't know on which device it will be executed.

As the name is the more important settings, it was at the root level. This have been defined when thinking to config format during a huddle. You can raise this question during today's standup !

> ### `options`

Reviewer: I'm sorry I didn't see your comment before standup today. So I am curious, could I conceivably request a devices based ONLY on one of the options values? Like, `deviceType: TABLET`? If so, I am not sure why these device request filters are not ALL at the devices root level?

> ### `options`

Author: @valentyn88 @alexplischke do you remember the reason those items where under `options` ?

> ### `options`

Reviewer 2: No particular strong reason either afaik. It was a choice between having it at the device level and under options. No one spoke out against having those fields under `options` and I faintly recall Kunal's preference for it. Simple as that 😄

> ### `options`

Reviewer: Fair enough.

> __Description__:  Device request options.

Reviewer: ```suggestion
__Description__:  Further specify certain device attributes within the pool of devices that match the `name` and `version` criteria.
```
Doe this make sense, Felix? Am I understanding the behavior correctly?

__Type__: *object*

#### `carrierConnectivity`

> __Description__: Request a device that is connected to cellular network.

Reviewer: ```suggestion
__Description__: Request that the matching device is also connected to a cellular network.
```

> __Type__:  *string*

Author: You're totally right !

__Example__:

```yaml
       carrierConnectivity: true

```

#### `deviceType`

> __Description__:  Request a device from a specific kind. Values:  `ANY`, `TABLET`, `PHONE`.

Reviewer: ```suggestion
__Description__:  Request that the matching device is a specific type of device. Values:  `ANY`, `TABLET`, or `PHONE`.
```

> __Type__:  *string*

Reviewer: Should this not be an enum, since the acceptable values are prescribed?

Author: Since it's a yaml file, enum is not really a thing. I don't know if we want `Type` to match the real yaml type or the expected implementation type. What's your point on this ?

Reviewer: I guess it doesn't really matter. I just always assume that any element for which there is a prescribed list of valid values is an enum. :)

Author: I think that's better to keep it as it. We could discuss with the rest of the team during standup today !

Reviewer: No need. I am fine with that decision. :)

__Example__:

```yaml

>         deviceTypecarrierConnectivity: TABLET

Reviewer: Is this value case sensitive?

Author: Yes it is. We may consider make `saucectl` to set everything uppercase to avoid weird behaviour to the user !

```

#### `private`

> __Description__: Request a device from a private pool.

Reviewer: ```suggestion
__Description__: Request that the matching device is from your organization's private pool.
```
